File: solutions/php/meetup/1/Meetup.php
<?php
declare(strict_types=1);

function meetup_day(int $year, int $month, string $which, string $weekday): DateTimeImmutable
{
    $date = new DateTimeImmutable("$year-$month-01");
    
    switch ($which) {
        case 'first':
            // First weekday of month
            return $date->modify("first $weekday of this month");
            
        case 'second':
            return $date->modify("first $weekday of this month")->modify('+1 week');
            
        case 'third':
            return $date->modify("first $weekday of this month")->modify('+2 weeks');
            
        case 'fourth':
            return $date->modify("first $weekday of this month")->modify('+3 weeks');
            
        case 'last':
            return $date->modify("last $weekday of this month");
            
        case 'teenth':
            // Days 13-19
            for ($day = 13; $day <= 19; $day++) {
                $candidate = new DateTimeImmutable("$year-$month-$day");
                if ($candidate->format('l') === $weekday) {
                    return $candidate;
                }
            }
            // Should never reach here
            throw new \Exception("No teenth $weekday found in $year-$month");
            
        default:
            throw new \InvalidArgumentException("Invalid 'which': $which");
    }
}
?>

File: solutions/ruby/meetup/1/meetup.rb
require 'date'

class Meetup
  def initialize(month, year)
    @month = month
    @year = year
  end

  def day(weekday, week)
    # 1. Determinamos los límites del rango de días a evaluar
    start_day, end_day = case week
                         when :first  then [1, 7]
                         when :second then [8, 14]
                         when :third  then [15, 21]
                         when :fourth then [22, 28]
                         when :teenth then [13, 19]
                         when :last
                           # Para :last, buscamos el último día del mes (día -1 en Date)
                           last_day_of_month = Date.new(@year, @month, -1).day
                           [last_day_of_month - 6, last_day_of_month]
                         end

    # 2. Generamos las fechas reales dentro de ese rango
    possible_dates = (start_day..end_day).map { |d| Date.new(@year, @month, d) }

    # 3. Filtramos la fecha que coincida con el método dinámico (ej. :monday? o :sunday?)
    # El método de consulta termina en '?', ej: :monday -> :monday?
    query_method = "#{weekday}?".to_sym

    possible_dates.find { |date| date.send(query_method) }
  end
end